### DISCUSSION TITLE
Next.js 14.11 canary - Your target environment does not appear to support 'async/await' (pdfjs-dist error)?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#61551] [Help]
### Summary

In trying to resolve https://github.com/vercel/next.js/discussions/61549, I upgraded to Next.js canary just now (14.1.1-canary.27), and now I'm getting a host of other errors. First was tailwind css directives not working, so I added a custom CSS rule and disabled the default CSS transpilation. But the main issue I'm facing now is my PDF UI tool is no longer working. I just had fully working locally a few hours ago using pdfjs-dist and Next.js 14 (but that first link error showed on `pnpm build` or deploy).

Now I am getting this in the browser (same code underneath, just updated to canary):

```
hot-reloader-client.js:242 ./node_modules/.pnpm/pdfjs-dist@4.0.379/node_modules/pdfjs-dist/build/pdf.mjs
The generated code contains 'async/await' because this module is using "topLevelAwait".
However, your target environment does not appear to support 'async/await'.
As a result, the code may not run as expected or may cause runtime errors.
processMessage @ hot-reloader-client.js:242
handler @ hot-reloader-client.js:411
pdf:1 Third-party cookie will be blocked. Learn more in the Issues tab.
pdf.mjs:17513 Warning: Setting up fake worker.
pdf.mjs:3733 Uncaught (in promise) Error: Setting up fake worker failed: "Failed to resolve module specifier "@swc/helpers/_/_class_private_field_get". Relative references must start with either "/", "./", or "../".".
    at eval (pdf.mjs:3733:36)
```

And in the terminal:

```
 ⚠ Port 3000 is in use, trying 3001 instead.
   ▲ Next.js 14.1.1-canary.27
   - Local:        http://localhost:3001
   - Environments: .env.local, .env

Warning: Built-in CSS support is being disabled due to custom CSS configuration being detected.
See here for more info: https://nextjs.org/docs/messages/built-in-css-disabled

Warning: Built-in CSS support is being disabled due to custom CSS configuration being detected.
See here for more info: https://nextjs.org/docs/messages/built-in-css-disabled

   automatically enabled Fast Refresh for 1 custom loader
 ✓ Ready in 2.4s
objc[71904]: Class GNotificationCenterDelegate is implemented in both /opt/homebrew/Cellar/glib/2.78.3/lib/libgio-2.0.0.dylib (0x11d7ac6b0) and /Users/lancepollard/start/termsurf/deck/site/task.surf/node_modules/.pnpm/sharp@0.32.6/node_modules/sharp/vendor/8.14.5/darwin-arm64v8/lib/libvips-cpp.42.dylib (0x13f6bda40). One of the two will be used. Which one is undefined.
 ○ Compiling /split/pdf ...
 ⚠ ./node_modules/.pnpm/pdfjs-dist@4.0.379/node_modules/pdfjs-dist/build/pdf.mjs
The generated code contains 'async/await' because this module is using "topLevelAwait".
However, your target environment does not appear to support 'async/await'.
As a result, the code may not run as expected or may cause runtime errors.

Import trace for requested module:
./node_modules/.pnpm/pdfjs-dist@4.0.379/node_modules/pdfjs-dist/build/pdf.mjs
./utilities/pdfjs.ts
./components/edit/EditPdf.tsx
./app/split/pdf/page.tsx
```

I lazy-load the `utilities/pdfjs.ts`, and that file contains:

```
import * as pdfjs from 'pdfjs-dist'

pdfjs.GlobalWorkerOptions.workerSrc = new URL(
  'pdfjs-dist/build/pdf.worker.min.mjs',
  import.meta.url,
).toString()

export default pdfjs
```

In my EditPDF page.tsx, I have:

```
const handleSelectFiles = (files: Array<File>) => {
  files.forEach(async file => {
    const arrayBuffer = await readFileAsArrayBuffer(file)
    const pdfjs = (await import('~/utilities/pdfjs')).default
    // do stuf with pdfjs
  })
```

So I think that should be fine? How do I resolve that `your target environment does not appear to support 'async/await'.` error?

Locally I am using Node v20.10.0.

**Comment by biomathcode** [comment]
@lancejpollard did you find any solution?

**Reply by lancejpollard** [reply]
Main thing I remember is I downgraded to `"pdfjs-dist": "3.11.174"`, and everything seems to work now.

**Comment by mmo80** [comment]
i got `pdfjs-dist ^4.0.379` (pdf.js) to work with `Next.js 14.1.4` for api route but does not work with Server Actions for some reason.

[https://github.com/mmo80/pdfja-nextjs-bug](https://github.com/mmo80/pdfja-nextjs-bug)

**Comment by enricoros** [comment]
I get the same issue with both pdfJS and tiktoken, when upgrading to `Next.js 14.2.3`:
```
./node_modules/pdfjs-dist/build/pdf.mjs
The generated code contains 'async/await' because this module is using "topLevelAwait".
However, your target environment does not appear to support 'async/await'.
As a result, the code may not run as expected or may cause runtime errors.
```

**Comment by am-unklown** [comment]
Probably its related to Edge runtime + middleware / instrumentation.ts / external libs.
Check #64792 (and specifically [this comment](https://github.com/vercel/next.js/issues/64792#issuecomment-2121526614)).

[Edge runtime](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes) is a vercel thing. Middlewares are restricted to use it. Not all node features are supported there (including top level await) and not all libs will work. There are discussions about allowing users to use nodejs runtime instead ([1](https://github.com/vercel/next.js/discussions/34179), [2](https://github.com/vercel/next.js/discussions/46722), [3](https://github.com/vercel/next.js/discussions/71727)). 

<details>
  <summary>Is not related to this issue but can be helpful to someone</summary>
(Sidenote: issue is not related to tsconfigs target probably. Not sure but webpack may not consider that target when builds your app at all.)

In my case I called func inside instrumentation.ts. That func uses prisma (and inits it if not yet) and produces log entries like yours (about not supported top level async await). Placing that func in app route (just for testing) does not produce log warns.

So i wrapped that function like this (according to [docs](https://nextjs.org/docs/app/building-your-application/optimizing/instrumentation)):
```
export async function register() {
  if (process.env.NEXT_RUNTIME === 'nodejs') {
    await import('./my-func-which-causes-warnings-about-top-level-await')
  }
}
```
And log warning dissapears. In case you face such problem in middleware I assume this will not work as only edge runtime available there.
</details>
